### PR TITLE
Improve session termination reason handling

### PR DESCRIPTION
**Pull request by vkatsuba**
Relate to https://github.com/travelping/ergw/pull/271

### CODE REVIEW

**Comment by vkatsuba** [comment]
> Is there anybody matching on 'inactivity_timeout' ?

The `inactivity_timeout` was change to `cp_inactivity_timeout` and `up_inactivity_timeout`.

**Comment by ebengt** [comment]
Yes, it was changed. So if anybody is matching on the old value it will not work anymore. Therefore the question.

File: src/ergw_aaa_static.erl [hunk 1]
```diff
@@ -87,8 +87,8 @@ handle_response(Procedure, #{'Result-Code' := Code} = Avps,
   when Code < 3000 ->
     {Session, Events} = to_session(Procedure, {Session0, Events0}, Avps),
     {ok, Session, Events, State};
-handle_response(_Procedure, #{'Result-Code' := Code}, Session, Events, State) ->
-    {{fail, Code}, Session, [stop | Events], State};
+handle_response(Procedure, #{'Result-Code' := Code}, Session, Events, State) ->
+    {{fail, Code}, Session, [{stop, Procedure} | Events], State};
```
Comment (RoadRunnr) [inline src/ergw_aaa_static.erl:91]: couldn't that lead to a situation where we return a `Reason` that has no mapping to a `Termination-Cause` ?

File: src/ergw_aaa_ro.erl [hunk 2]
```diff
@@ -323,7 +323,7 @@ handle_cca(['CCA' | #{'Result-Code' := Code} = Avps],
     {ok, Session, Events, State};
 handle_cca([Answer | #{'Result-Code' := Code}], Session, Events, _Opts, State)
   when Answer =:= 'CCA'; Answer =:= 'answer-message' ->
-    {{fail, Code}, Session, [stop | Events], State};
+    {{fail, Code}, Session, [{stop, Answer} | Events], State};
```
Comment (RoadRunnr) [inline src/ergw_aaa_ro.erl:326]: Knowing which procedure cause the stop is helpful, but the failure code would also make sense.

**Review by RoadRunnr — CHANGES_REQUESTED** [review]
The `Reason` returned in a `stop` event does not seem consistent across all places nor sensible in some instances.

**Comment by vkatsuba** [comment]
> Yes, it was changed. So if anybody is matching on the old value it will not work anymore. Therefore the question.

It is not an issue, because in different versions will be use different default mapping and this info will be provided in news for release, plus this can be re-configured(just need update configuration of `ergw_aaa`). The logic for support of new default mapping in ergw for new release in progress.

File: src/ergw_aaa_static.erl (hunk 1 above)
Comment (vkatsuba) [reply src/ergw_aaa_static.erl:91]: Yes, the reason will not be mapped of termination cause, the main goal of this changes - return reason of stop session for handle it in `ergw`.

**Comment by vkatsuba** [comment]
> The `Reason` returned in a `stop` event does not seem consistent across all places nor sensible in some instances.

Do you have any other approaches or visions, how we can return termination reasons from `ergw_aaa` to `ergw` and handle it?

**Comment by RoadRunnr** [comment]
> > The `Reason` returned in a `stop` event does not seem consistent across all places nor sensible in some instances.
> 
> Do you have any other approaches or visions, how we can return termination reasons from `ergw_aaa` to `ergw` and handle it?

I have not thought about that in depth, but I was under impression that @fholzhauser had?

In any case, returning something like `answer-message' (or other generic terms) as stop cause is obviously not helpful. Using increasingly complex atoms is also not the right way. It might be better to use tuples there.

**Comment by fholzhauser** [comment]
Actually I didn't really consider pushing received diameter error results towards the ERGW. That indeed adds another "dimension" to the possible responses to handle. I'm also not sure it adds a lot of value.

Refining termination causes served 2 purposes. 
- Observability : to show the different reasons why sessions are terminated in metrics. I think for this just using e.g. `aaa_peer_reject` (or similar) for any errors received from a remote AAA peer would do. The diameter metrics are quite good and it'd indicate where to look next. Having the handler module name could be interesting though.
- Feedback to external systems. Probably it'd mostly be other AAA peers, as on the other interfaces we can't do much with reasons.

So I'd probably go with a tuple with `{Handler, Error}` where indeed `Error` would be atoms for events that we generate (e.g. timeout, no connection, rate limit etc. situations) and a generic one e,g, `aaa_peer_reject` for any error responses. 

Would this make sense ?

**Review by fholzhauser — APPROVED** [review]

**Pull request merged**